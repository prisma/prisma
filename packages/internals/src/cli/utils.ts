import Arg from 'arg'
import dedent from 'strip-indent'

/**
 * format
 */
export function format(input = ''): string {
  return dedent(input).trimRight() + '\n'
}

/**
 * Wrap arg to return an error instead of throwing
 */
export function arg<T extends Arg.Spec>(
  argv: string[],
  spec: T,
  stopAtPositional = true,
  permissive = false,
): Arg.Result<T> | Error {
  try {
    return Arg(spec, { argv, stopAtPositional, permissive })
  } catch (e: any) {
    return e
  }
}

/**
 * Check if result is an error
 */
export function isError(result: any): result is Error {
  return result instanceof Error
}
