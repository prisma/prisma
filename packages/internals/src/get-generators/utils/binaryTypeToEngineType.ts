import { BinaryType } from '@prisma/fetch-engine'
import type { EngineType } from '@prisma/generator-helper'

export function binaryTypeToEngineType(binaryType: string): EngineType {
  if (binaryType === BinaryType.introspectionEngine) {
    return 'introspectionEngine'
  }

  if (binaryType === BinaryType.migrationEngine) {
    return 'migrationEngine'
  }
  if (binaryType === BinaryType.libqueryEngine) {
    return 'libqueryEngine'
  }
  if (binaryType === BinaryType.queryEngine) {
    return 'queryEngine'
  }

  if (binaryType === BinaryType.prismaFmt) {
    return 'prismaFmt'
  }

  throw new Error(`Could not convert binary type ${binaryType}`)
}
