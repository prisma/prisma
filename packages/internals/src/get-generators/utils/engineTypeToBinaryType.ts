import { BinaryType } from '@prisma/fetch-engine'
import type { EngineType } from '@prisma/generator-helper'

export function engineTypeToBinaryType(engineType: EngineType): BinaryType {
  if (engineType === 'introspectionEngine') {
    return BinaryType.introspectionEngine
  }

  if (engineType === 'migrationEngine') {
    return BinaryType.migrationEngine
  }

  if (engineType === 'queryEngine') {
    return BinaryType.queryEngine
  }
  if (engineType === 'libqueryEngine') {
    return BinaryType.libqueryEngine
  }
  if (engineType === 'prismaFmt') {
    return BinaryType.prismaFmt
  }

  throw new Error(`Could not convert engine type ${engineType}`)
}
