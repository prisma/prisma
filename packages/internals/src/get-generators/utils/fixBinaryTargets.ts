import { BinaryTargetsEnvValue } from '@prisma/generator-helper'
import { BinaryTarget } from '@prisma/get-platform'

// eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
function transformBinaryTargetToEnvValue(binaryTarget: BinaryTarget | string): BinaryTargetsEnvValue {
  return { fromEnvVar: null, value: binaryTarget }
}

export function fixBinaryTargets(
  schemaBinaryTargets: BinaryTargetsEnvValue[],
  // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
  runtimeBinaryTarget: BinaryTarget | string,
): BinaryTargetsEnvValue[] {
  schemaBinaryTargets = schemaBinaryTargets || []

  if (!schemaBinaryTargets.find((object) => object.native === true)) {
    return [transformBinaryTargetToEnvValue('native'), ...schemaBinaryTargets]
  }

  return [...schemaBinaryTargets, transformBinaryTargetToEnvValue(runtimeBinaryTarget)]
}
