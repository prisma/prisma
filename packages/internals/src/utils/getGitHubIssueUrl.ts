import { getPlatform } from '@prisma/get-platform'
import isWindows from 'is-windows'
import isWSL from 'is-wsl'
import newGitHubIssueUrl from 'new-github-issue-url'
import open from 'open'
import prompt from 'prompts'
import stripAnsi from 'strip-ansi'
import { match } from 'ts-pattern'

export function getGitHubIssueUrl({
  title,
  user = 'prisma',
  repo = 'prisma',
  template = 'bug_report.md',
  body,
}: {
  title: string
  user?: string
  repo?: string
  template?: string
  body?: string
}): string {
  return newGitHubIssueUrl({
    user,
    repo,
    template,
    title,
    body,
  })
}

interface IssueOptions {
  error: any
  cliVersion: string
  enginesVersion: string
  command: string
  prompt: Boolean
  title?: string
  reportId?: number
}

export async function wouldYouLikeToCreateANewIssue(options: IssueOptions) {
  const shouldCreateNewIssue = await match(options.prompt)
    .with(true, async () => {
      const createNewIssueResponse = await prompt({
        type: 'select',
        name: 'value',
        message: 'Would you like to create a GitHub issue?',
        initial: 0,
        choices: [
          {
            title: 'Yes',
            value: true,
            description: `Create a new GitHub issue`,
          },
          {
            title: 'No',
            value: false,
            description: `Don't create a new GitHub issue`,
          },
        ],
      })
      // @types/prompts is broken, and doesn't infer the types of the `value` prop,
      // thus we explicitly cast it to boolean.
      return Boolean(createNewIssueResponse.value)
    })
    .otherwise(() => Promise.resolve(true))

  if (shouldCreateNewIssue) {
    const platform = await getPlatform()

    const url = getGitHubIssueUrl({
      title: options.title ?? '',
      body: issueTemplate(platform, options),
    })

    /**
     * This is a workaround for a quirky `open` behavior.
     * - `await open(url)` correctly opens the browser, returns the control to Node.js and let it exit the process normally.
     * - `await open(url, { wait: true })` achieves the same on Windows and WSL. Without `{ wait: true }`, `open` would not
     *   return the control to the Node.js process, which would result in the prisma cli being stuck.
     */
    const shouldOpenWait = isWindows() || isWSL
    await open(url, { wait: shouldOpenWait })
  } else {
    // Return SIGINT exit code to signal that the process was cancelled.
    process.exit(130)
  }
}

const issueTemplate = (platform: string, options: IssueOptions) => {
  return stripAnsi(`
Hi Prisma Team! The following command just crashed.
${options.reportId ? `The report Id is: ${options.reportId}` : ''}

## Command

\`${options.command}\`

## Versions
      
| Name        | Version            |
|-------------|--------------------|
| Platform    | ${platform.padEnd(19)}| 
| Node        | ${process.version.padEnd(19)}| 
| Prisma CLI  | ${options.cliVersion.padEnd(19)}| 
| Engine      | ${options.enginesVersion.padEnd(19)}| 

## Error
\`\`\`
${options.error}
\`\`\`
`)
}
