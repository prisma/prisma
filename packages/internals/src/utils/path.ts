import path from 'path'

/**
 * Normalize `filePath` to use forward slashes as a separator. `filePath` is
 * treated as a path specific to the current platform, so backslashes will only
 * be replaced with forward slashes on Windows. On other platforms, where a
 * backslash is a valid filename character, it will be treated as such and will
 * not be replaced.
 */
export function pathToPosix(filePath: string): string {
  if (path.sep === path.posix.sep) {
    return filePath
  }
  return filePath.split(path.sep).join(path.posix.sep)
}

/**
 * Returns the longest common path ancestor of the two paths (which may also be equal to one or both of them).
 * If they don't share a common ancestor (which may be the case on Windows if both paths are on different disks),
 * returns `undefined`.
 */
export function longestCommonPathPrefix(pathA: string, pathB: string): string | undefined {
  if (!path.isAbsolute(pathA) || !path.isAbsolute(pathB)) {
    throw new Error('longestCommonPathPrefix expects absolute paths')
  }

  if (process.platform === 'win32' && (pathA.startsWith('\\\\') || pathB.startsWith('\\\\'))) {
    // Make both paths namespaced if at least one of them is.
    pathA = path.toNamespacedPath(pathA)
    pathB = path.toNamespacedPath(pathB)
  }

  const commonPrefix = longestCommonPrefix(pathA.split(path.sep), pathB.split(path.sep)).join(path.sep)

  if (commonPrefix === '') {
    return process.platform === 'win32' ? undefined : '/'
  }

  if (process.platform === 'win32' && ['\\', '\\\\?', '\\\\.'].includes(commonPrefix)) {
    return undefined
  }

  if (process.platform === 'win32' && commonPrefix.endsWith(':')) {
    // Disk specifier without a backslash at the end is not an absolute path on windows,
    // it refers to current working directory on that disk.
    return commonPrefix + '\\'
  }

  return commonPrefix
}

function longestCommonPrefix<T>(sequenceA: T[], sequenceB: T[]): T[] {
  const maxLen = Math.min(sequenceA.length, sequenceB.length)
  let sliceLen = 0

  while (sliceLen <= maxLen && sequenceA[sliceLen] === sequenceB[sliceLen]) {
    sliceLen++
  }

  return sequenceA.slice(0, sliceLen)
}
