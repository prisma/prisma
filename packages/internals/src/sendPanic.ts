import Debug from '@prisma/debug'
import { getPlatform } from '@prisma/get-platform'
import archiver from 'archiver'
import * as checkpoint from 'checkpoint-client'
import fs from 'fs'
import globby from 'globby'
import os from 'os'
import path from 'path'
import stripAnsi from 'strip-ansi'
import tmp from 'tmp'
import { match, P } from 'ts-pattern'

import { createErrorReport, ErrorKind, makeErrorReportCompleted, uploadZip } from './errorReporting'
import type { RustPanic } from './panic'
import { ErrorArea } from './panic'
import { mapScalarValues, maskSchema } from './utils/maskSchema'

const debug = Debug('prisma:sendPanic')
// cleanup the temporary files even when an uncaught exception occurs
tmp.setGracefulCleanup()

type SendPanic = {
  error: RustPanic
  cliVersion: string
  enginesVersion: string

  // retrieve the database version for the given schema or url, without throwing any error
  getDatabaseVersionSafe: (schemaOrUrl: string) => Promise<string | undefined>
}
export async function sendPanic({
  error,
  cliVersion,
  enginesVersion,
  getDatabaseVersionSafe,
}: SendPanic): Promise<number> {
  try {
    const schema: string | undefined = match(error)
      .with({ schemaPath: P.when((schemaPath) => Boolean(schemaPath)) }, (err) => {
        return fs.readFileSync(err.schemaPath, 'utf-8')
      })
      .with({ schema: P.when((schema) => Boolean(schema)) }, (err) => err.schema)
      .otherwise(() => undefined)

    const maskedSchema: string | undefined = schema ? maskSchema(schema) : undefined

    let dbVersion: string | undefined
    // For a SQLite datasource like `url = "file:dev.db"` only schema will be defined
    const schemaOrUrl = schema || error.introspectionUrl
    if (error.area === ErrorArea.LIFT_CLI && schemaOrUrl) {
      dbVersion = await getDatabaseVersionSafe(schemaOrUrl)
    }

    const migrateRequest = error.request
      ? JSON.stringify(
          mapScalarValues(error.request, (value) => {
            if (typeof value === 'string') {
              return maskSchema(value)
            }
            return value
          }),
        )
      : undefined

    const params = {
      area: error.area,
      kind: ErrorKind.RUST_PANIC,
      cliVersion,
      binaryVersion: enginesVersion,
      command: getCommand(),
      jsStackTrace: stripAnsi(error.stack || error.message),
      rustStackTrace: error.rustStack,
      operatingSystem: `${os.arch()} ${os.platform()} ${os.release()}`,
      platform: await getPlatform(),
      liftRequest: migrateRequest,
      schemaFile: maskedSchema,
      fingerprint: await checkpoint.getSignature(),
      sqlDump: undefined,
      dbVersion: dbVersion,
    }

    const signedUrl = await createErrorReport(params)

    if (error.schemaPath) {
      const zip = await makeErrorZip(error)
      await uploadZip(zip, signedUrl)
    }

    const id = await makeErrorReportCompleted(signedUrl)
    return id
  } catch (e) {
    debug(e)
    throw e
  }
}

function getCommand(): string {
  // don't send url
  if (process.argv[2] === 'introspect') {
    return 'introspect'
  } else if (process.argv[2] === 'db' && process.argv[3] === 'pull') {
    return 'db pull'
  }
  return process.argv.slice(2).join(' ')
}

async function makeErrorZip(error: RustPanic): Promise<Buffer> {
  if (!error.schemaPath) {
    throw new Error(`Can't make zip without schema path`)
  }
  const schemaDir = path.dirname(error.schemaPath)
  const tmpFileObj = tmp.fileSync()
  const outputFile = fs.createWriteStream(tmpFileObj.name)
  const zip = archiver('zip', { zlib: { level: 9 } })

  zip.pipe(outputFile)

  // add schema file
  // Note: the following reads `error.schemaPath` for the second time, we could just re-use
  // `maskedSchema` from the `sendPanic` function's scope.
  const schemaFile = maskSchema(fs.readFileSync(error.schemaPath, 'utf-8'))
  zip.append(schemaFile, { name: path.basename(error.schemaPath) })

  if (fs.existsSync(schemaDir)) {
    const filePaths = await globby('migrations/**/*', {
      // globby doesn't have it in its types but it's part of mrmlnc/fast-glob
      // @ts-ignore
      cwd: schemaDir,
    })

    for (const filePath of filePaths) {
      let file = fs.readFileSync(path.resolve(schemaDir, filePath), 'utf-8')
      if (filePath.endsWith('schema.prisma') || filePath.endsWith(path.basename(error.schemaPath))) {
        // Remove credentials from schema datasource url
        file = maskSchema(file)
      }
      zip.append(file, { name: path.basename(filePath) })
    }
  }

  zip.finalize()

  return new Promise((resolve, reject) => {
    outputFile.on('close', () => {
      const buffer = fs.readFileSync(tmpFileObj.name)
      resolve(buffer)
    })

    zip.on('error', (err) => {
      reject(err)
    })
  })
}
