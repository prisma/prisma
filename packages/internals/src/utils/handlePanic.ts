import chalk from 'chalk'
import prompt from 'prompts'

import type { RustPanic } from '../panic'
import { sendPanic } from '../sendPanic'
import { canPrompt } from './canPrompt'
import { wouldYouLikeToCreateANewIssue } from './getGitHubIssueUrl'
import { link } from './link'

type HandlePanic = {
  error: RustPanic
  cliVersion: string
  enginesVersion: string
  command: string

  // retrieve the database version for the given schema or url, without throwing any error
  getDatabaseVersionSafe: (schemaOrUrl: string) => Promise<string | undefined>
}

export async function handlePanic(args: HandlePanic): Promise<void> {
  if (!canPrompt()) {
    throw args.error
  }

  await panicDialog(args)
}

type PanicDialog = HandlePanic

async function panicDialog({
  error,
  cliVersion,
  enginesVersion,
  command,
  getDatabaseVersionSafe,
}: PanicDialog): Promise<void> {
  const errorMessage = error.message.split('\n').slice(0, Math.max(20, process.stdout.rows)).join('\n')

  console.log(`${chalk.red('Oops, an unexpected error occurred!')}
${chalk.red(errorMessage)}

${chalk.bold('Please help us improve Prisma by submitting an error report.')}
${chalk.bold('Error reports never contain personal or other sensitive information.')}
${chalk.dim(`Learn more: ${link('https://pris.ly/d/telemetry')}`)}
`)

  const { value: shouldSubmitReport } = await prompt({
    type: 'select',
    name: 'value',
    message: 'Submit error report',
    initial: 0,
    choices: [
      {
        title: 'Yes',
        value: true,
        description: `Send error report once`,
      },
      {
        title: 'No',
        value: false,
        description: `Don't send error report`,
      },
    ],
  })

  if (shouldSubmitReport) {
    try {
      console.log('Submitting...')
      const reportId = await sendPanic({ error, cliVersion, enginesVersion, getDatabaseVersionSafe })
      console.log(`\n${chalk.bold(`We successfully received the error report id: ${reportId}`)}`)
      console.log(`\n${chalk.bold('Thanks a lot for your help! üôè')}`)
    } catch (error) {
      const reportFailedMessage = `${chalk.bold.red('Oops. We could not send the error report.')}`
      console.log(reportFailedMessage)
      console.error(`${chalk.gray('Error report submission failed due to: ')}`, error)
    }
  }

  await wouldYouLikeToCreateANewIssue({
    prompt: !shouldSubmitReport,
    error,
    cliVersion,
    enginesVersion,
    command,
  })

  // Signal that there was an error
  process.exit(1)
}
