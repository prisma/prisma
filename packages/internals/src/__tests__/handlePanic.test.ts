import { ensureDir } from 'fs-extra'
import { stdin } from 'mock-stdin'
import { join, resolve } from 'path'
import prompt from 'prompts'
import stripAnsi from 'strip-ansi'
import tempy from 'tempy'

import { ErrorArea, jestConsoleContext, jestContext, RustPanic } from '..'
import * as sendPanicUtils from '../sendPanic'
import * as githubUtils from '../utils/getGitHubIssueUrl'
import { handlePanic } from '../utils/handlePanic'

const keys = {
  up: '\x1B\x5B\x41',
  down: '\x1B\x5B\x42',
  enter: '\x0D',
  space: '\x20',
}
const sendKeystrokes = async (io) => {
  io.send(keys.down)
  io.send(keys.enter)
  await delay(10)
}
// helper function for timing
const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms)) // Mock stdin so we can send messages to the CLI

const testRootDir = tempy.directory()

// eslint-disable-next-line @typescript-eslint/unbound-method
const oldProcessCwd = process.cwd

const ctx = jestContext.new().add(jestConsoleContext()).assemble()

describe('handlePanic', () => {
  // testing with env https://stackoverflow.com/a/48042799/1345244
  const OLD_ENV = process.env

  // mock for retrieving the database version
  const getDatabaseVersionSafe = () => Promise.resolve(undefined)

  beforeEach(async () => {
    jest.resetModules() // most important - it clears the cache
    process.env = { ...OLD_ENV, GITHUB_ACTIONS: 'true' } // make a copy and simulate CI environment
    process.cwd = () => testRootDir
    await ensureDir(testRootDir)
  })
  afterEach(() => {
    process.cwd = oldProcessCwd
    // await del(testRootDir, { force: true }) // Need force: true because `del` does not delete dirs outside the CWD
  })
  let io
  beforeAll(() => (io = stdin()))
  afterAll(() => {
    process.env = OLD_ENV // restore old env
    io.restore()
  })

  const error = new RustPanic(
    'Some error message!',
    '',
    undefined,
    ErrorArea.LIFT_CLI,
    resolve(join('fixtures', 'blog', 'prisma', 'schema.prisma')),
  )
  const packageJsonVersion = '0.0.0'
  const enginesVersion = '734ab53bd8e2cadf18b8b71cb53bf2d2bed46517'
  const command = 'something-test'

  // Only works locally (not in CI)
  it.skip('ask to submit the panic error in interactive mode', async () => {
    const oldConsoleLog = console.log
    const logs: string[] = []
    console.log = (...args) => {
      logs.push(...args)
    }

    setTimeout(() => sendKeystrokes(io).then(), 5)

    try {
      await handlePanic({
        error,
        cliVersion: packageJsonVersion,
        enginesVersion: enginesVersion,
        command,
        getDatabaseVersionSafe,
      })
    } catch (e) {
      /* eslint-disable-next-line @typescript-eslint/no-unsafe-member-access */
      expect(stripAnsi(e.message)).toMatchSnapshot()
    }

    console.log = oldConsoleLog
    expect(stripAnsi(logs.join('\n'))).toMatchSnapshot()
  })

  it('no interactive mode in CI', async () => {
    try {
      await handlePanic({
        error,
        cliVersion: packageJsonVersion,
        enginesVersion,
        command,
        getDatabaseVersionSafe,
      })
    } catch (error) {
      error.schemaPath = 'Some Schema Path'
      expect(error).toMatchInlineSnapshot(`[Error: Some error message!]`)
      expect(JSON.stringify(error)).toMatchInlineSnapshot(
        `"{"__typename":"RustPanic","rustStack":"","area":"LIFT_CLI","schemaPath":"Some Schema Path"}"`,
      )
    }
  })

  it('when sendPanic fails, the user should be alerted by a reportFailedMessage', async () => {
    const cliVersion = 'test-cli-version'
    const enginesVersion = 'test-engine-version'
    const rustStackTrace = 'test-rustStack'
    const command = 'test-command'
    const sendPanicTag = 'send-panic-failed'

    const mockExit = jest.spyOn(process, 'exit').mockImplementation()

    const spySendPanic = jest
      .spyOn(sendPanicUtils, 'sendPanic')
      .mockImplementation(() => Promise.reject(new Error(sendPanicTag)))
      .mockName('mock-sendPanic')

    const spyWouldYouLikeToCreateANewIssue = jest
      .spyOn(githubUtils, 'wouldYouLikeToCreateANewIssue')
      .mockImplementation(() => Promise.resolve())
      .mockName('mock-wouldYouLikeToCreateANewIssue')

    const rustPanic = new RustPanic(
      'test-message',
      rustStackTrace,
      'test-request',
      ErrorArea.LIFT_CLI, // area
      undefined, // schemaPath
      undefined, // schema
      undefined, // introspectionUrl
    )

    prompt.inject(['y']) // submit report
    await handlePanic({
      error: rustPanic,
      cliVersion,
      enginesVersion,
      command,
      getDatabaseVersionSafe,
    })

    expect(spySendPanic).toHaveBeenCalledTimes(1)
    expect(spyWouldYouLikeToCreateANewIssue).toHaveBeenCalledTimes(1)
    expect(stripAnsi(ctx.mocked['console.log'].mock.calls.join('\n'))).toMatchSnapshot()
    expect(stripAnsi(ctx.mocked['console.error'].mock.calls.join('\n'))).toMatch(
      new RegExp(`^Error report submission failed due to:?`),
    )
    expect(mockExit).toHaveBeenCalledWith(1)
    spySendPanic.mockRestore()
  })
})
