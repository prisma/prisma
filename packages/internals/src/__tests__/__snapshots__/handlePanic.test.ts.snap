// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handlePanic when sendPanic fails, the user should be alerted by a reportFailedMessage 1`] = `
"Oops, an unexpected error occurred!


Please help us improve Prisma by submitting an error report.
Error reports never contain personal or other sensitive information.
Learn more: https://pris.ly/d/telemetry

Submitting...
Oops. We could not send the error report."
`;
